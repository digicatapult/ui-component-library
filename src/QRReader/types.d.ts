import type { QRCode } from 'jsqr'

export type DecodeRequest = {
  data: Uint8ClampedArray
  width: number
  height: number
}

export type DecodeResponse =
  | {
      type: 'success'
      code: QRCode
    }
  | {
      type: 'failure'
    }
