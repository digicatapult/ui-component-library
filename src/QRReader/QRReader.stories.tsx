import { Meta, StoryObj } from '@storybook/react'
import { action } from '@storybook/addon-actions'

import QRReader from './index.js'
import { styled } from 'styled-components'
import React from 'react'

export default {
  title: 'Components/QRReader',
  component: QRReader,
  argTypes: {
    showViewfinder: {
      controlType: 'boolean',
      defaultValue: true,
      description: 'Whether to show the viewfinder',
      type: 'boolean',
    },
  },
} as Meta<typeof QRReader>

export const Default: StoryObj<typeof QRReader> = {
  render: (props) => {
    return <QRStyled {...props} />
  },
  args: {
    onResult: action('result'),
    showViewfinder: true,
  },
}

const QRStyled = styled(QRReader)`
  width: 400px;
`
