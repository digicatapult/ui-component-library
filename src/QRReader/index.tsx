import { styled } from 'styled-components'
import type { DecodeRequest, DecodeResponse } from './types'

import React, { useState, useEffect, useRef, useCallback } from 'react'

import { createWorker } from './worker.util.js'
import { getUserMedia } from './userMedia.util.js'

export interface QRReaderProps {
  onResult: (result: string) => void
  scanDelay?: number
  viewFinderColor?: string
  viewFinderVariant?: 'viewfinder-cross-med' | 'viewfinder-none'
  className?: string
}

const delay = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms))

const QRReader: React.FC<QRReaderProps> = ({
  onResult,
  scanDelay = 500,
  viewFinderColor = '#55555580',
  viewFinderVariant = 'viewfinder-cross-med',
  className,
}) => {
  const [result, setResult] = useState<string | null>(null)
  const [error, setError] = useState<string | null>(null)
  const videoRef = useRef<HTMLVideoElement>(null)
  const canvasRef = useRef<HTMLCanvasElement>(null)
  const workerRef = useRef<Worker | null>(null)

  // kick off the webworker that will process qr decode requests
  useEffect(() => {
    const worker = createWorker()
    workerRef.current = worker

    return () => {
      worker.terminate()
    }
  }, [])

  // hook to initialise the media loop
  useEffect(() => {
    const constraints = { video: { facingMode: 'environment' } }
    let stream: MediaStream | null = null

    getUserMedia(constraints)
      .then((mediaStream) => {
        const video = videoRef.current
        if (!video) {
          // this should never happen
          throw new Error('Video was not initialised as expected')
        }

        video.srcObject = mediaStream
        video.play()
        stream = mediaStream
      })
      .catch((err) => {
        // TODO: trigger an error state and render a fallback
        console.error('Error accessing webcam:', err)
        setError('Error accessing webcam')
      })

    return () => {
      if (!stream) {
        return
      }
      const tracks = stream.getTracks()
      tracks.forEach((track) => {
        track.stop()
      })
    }
  }, [setError])

  const parseCurrentFrame = useCallback(async () => {
    const canvas = canvasRef.current
    const video = videoRef.current
    const worker = workerRef.current

    if (
      !canvas ||
      !video ||
      !worker ||
      video.readyState !== video.HAVE_ENOUGH_DATA
    ) {
      return null
    }

    const ctx = canvas.getContext('2d')
    if (!ctx) {
      return null
    }

    canvas.height = video.videoHeight * 0.8
    canvas.width = video.videoWidth * 0.8
    ctx.drawImage(
      video,
      0.1 * video.videoWidth,
      0.1 * video.videoHeight,
      video.videoWidth * 0.8,
      video.videoHeight * 0.8,
      0,
      0,
      canvas.width,
      canvas.height,
    )

    const imageData = ctx.getImageData(0, 0, canvas.width, canvas.height)

    const message: DecodeRequest = {
      data: imageData.data,
      width: imageData.width,
      height: imageData.height,
    }

    const result = await new Promise<string | null>((resolve) => {
      worker.onmessage = ({ data }: { data: DecodeResponse }) => {
        if (data.type === 'success') {
          const code = data.code
          resolve(code.data)
        }
        resolve(null)
      }

      worker.postMessage(message)
    })
    return result
  }, [])

  // start the decoding loop. Note we store the result in local state so that we don't need to restart the loop just because the onResult handler changes
  useEffect(() => {
    let stop = false
    const loop = async () => {
      while (true) {
        if (stop) {
          return
        }
        const result = await parseCurrentFrame()
        if (stop) {
          return
        }
        setResult(result)
        await delay(scanDelay)
      }
    }
    loop()

    return () => {
      stop = true
    }
  }, [setResult, scanDelay, parseCurrentFrame])

  // Update the parent on result changes whenever they happen
  useEffect(() => {
    if (result) {
      onResult(result)
    }
  }, [result, onResult])

  return error ? (
    <div className={className}>{error}</div>
  ) : (
    <Wrapper
      className={className}
      showViewfinder={viewFinderVariant === 'viewfinder-cross-med'}
      viewFinderColor={viewFinderColor}
    >
      <Video ref={videoRef} muted={true} />
      <canvas ref={canvasRef} hidden={true} />
    </Wrapper>
  )
}

const Wrapper = styled.div<{
  showViewfinder: boolean
  viewFinderColor: string
}>`
  width: 100%;
  position: relative;

  ${({ showViewfinder, viewFinderColor }) =>
    showViewfinder
      ? `
  &:before {
    content: '';
    position: absolute;
    display: inline-block;
    top: 10%;
    left: 10%;
    right: 10%;
    bottom: 10%;
    opacity: 0.2;
    background:
      linear-gradient(to right, ${viewFinderColor} 4px, transparent 4px) 0 0,
      linear-gradient(to right, ${viewFinderColor} 4px, transparent 4px) 0 100%,
      linear-gradient(to left, ${viewFinderColor} 4px, transparent 4px) 100% 0,
      linear-gradient(to left, ${viewFinderColor} 4px, transparent 4px) 100% 100%,
      linear-gradient(to bottom, ${viewFinderColor} 4px, transparent 4px) 0 0,
      linear-gradient(to bottom, ${viewFinderColor} 4px, transparent 4px) 100% 0,
      linear-gradient(to top, ${viewFinderColor} 4px, transparent 4px) 0 100%,
      linear-gradient(to top, ${viewFinderColor} 4px, transparent 4px) 100% 100%;

    background-repeat: no-repeat;
    background-size: 20px 20px;
  }

  &:after {
    content: '';
    position: absolute;
    display: inline-block;
    top: 10%;
    left: 10%;
    right: 10%;
    bottom: 10%;
    opacity: 0.2;
    background:
      linear-gradient(to right, ${viewFinderColor} 4px, transparent 4px)
        calc(50% + 30px - 2px) 50%,
      linear-gradient(to bottom, ${viewFinderColor} 4px, transparent 4px) 50%
        calc(50% + 30px - 2px);

    background-repeat: no-repeat;
    background-size: 60px 60px;
  }`
      : ''}
`

const Video = styled.video`
  width: 100%;
`

export default QRReader
