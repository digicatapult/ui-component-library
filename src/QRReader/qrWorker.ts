import type { DecodeRequest, DecodeResponse } from './types'

import jsQR from 'jsqr'

// note this is intentionally defensive to ensure any message received
// is responded to. jsQR *might* throw
self.onmessage = ({ data }: { data: DecodeRequest }) => {
  let response: DecodeResponse
  try {
    const code = jsQR(data.data, data.width, data.height)
    response = code
      ? {
          type: 'success',
          code
        }
      : {
          type: 'failure'
        }
  } catch (err) {
    response = { type: 'failure' }
  }

  self.postMessage(response)
}
