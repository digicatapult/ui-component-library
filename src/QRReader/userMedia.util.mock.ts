export const getUserMedia = jest.fn().mockImplementation(
  () =>
    new Promise((r) =>
      setTimeout(r, 0, {
        getTracks: () => []
      })
    )
)
