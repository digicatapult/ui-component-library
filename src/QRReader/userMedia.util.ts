export const getUserMedia = (constraints: MediaStreamConstraints) => navigator.mediaDevices.getUserMedia(constraints)
