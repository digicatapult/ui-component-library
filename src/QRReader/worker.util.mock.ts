export const createWorker = () => ({
  terminate: jest.fn().mockReturnValue(undefined),
  postMessage: jest.fn().mockReturnValue(undefined)
})
