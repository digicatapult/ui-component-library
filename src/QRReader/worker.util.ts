export const createWorker = () => new Worker(new URL('./qrWorker.js', import.meta.url))
