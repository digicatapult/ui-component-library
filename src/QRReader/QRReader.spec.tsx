import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'

import QRReader from './index.js'

describe('QRReader', () => {
  test('default', () => {
    const tree = renderer.create(<QRReader onResult={() => {}} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('no viewfinder', () => {
    const tree = renderer
      .create(
        <QRReader onResult={() => {}} viewFinderVariant="viewfinder-none" />,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('modified viewfinder color', () => {
    const tree = renderer
      .create(<QRReader onResult={() => {}} viewFinderColor="red" />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
