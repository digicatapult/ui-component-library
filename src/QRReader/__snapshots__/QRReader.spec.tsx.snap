// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QRReader default 1`] = `
.c0 {
  width: 100%;
  position: relative;
}

.c0:before {
  content: '';
  position: absolute;
  display: inline-block;
  top: 10%;
  left: 10%;
  right: 10%;
  bottom: 10%;
  opacity: 0.2;
  background: linear-gradient(to right, #55555580 4px, transparent 4px) 0 0,linear-gradient(to right, #55555580 4px, transparent 4px) 0 100%,linear-gradient(to left, #55555580 4px, transparent 4px) 100% 0,linear-gradient(to left, #55555580 4px, transparent 4px) 100% 100%,linear-gradient(to bottom, #55555580 4px, transparent 4px) 0 0,linear-gradient(to bottom, #55555580 4px, transparent 4px) 100% 0,linear-gradient(to top, #55555580 4px, transparent 4px) 0 100%,linear-gradient(to top, #55555580 4px, transparent 4px) 100% 100%;
  background-repeat: no-repeat;
  background-size: 20px 20px;
}

.c0:after {
  content: '';
  position: absolute;
  display: inline-block;
  top: 10%;
  left: 10%;
  right: 10%;
  bottom: 10%;
  opacity: 0.2;
  background: linear-gradient(to right, #55555580 4px, transparent 4px) calc(50% + 30px - 2px) 50%,linear-gradient(to bottom, #55555580 4px, transparent 4px) 50% calc(50% + 30px - 2px);
  background-repeat: no-repeat;
  background-size: 60px 60px;
}

.c1 {
  width: 100%;
}

<div
  className="c0"
>
  <video
    className="c1"
    muted={true}
  />
  <canvas
    hidden={true}
  />
</div>
`;

exports[`QRReader mirrored 1`] = `
.c0 {
  width: 100%;
  position: relative;
}

.c0:before {
  content: '';
  position: absolute;
  display: inline-block;
  top: 10%;
  left: 10%;
  right: 10%;
  bottom: 10%;
  opacity: 0.2;
  background: linear-gradient(to right, #55555580 4px, transparent 4px) 0 0,linear-gradient(to right, #55555580 4px, transparent 4px) 0 100%,linear-gradient(to left, #55555580 4px, transparent 4px) 100% 0,linear-gradient(to left, #55555580 4px, transparent 4px) 100% 100%,linear-gradient(to bottom, #55555580 4px, transparent 4px) 0 0,linear-gradient(to bottom, #55555580 4px, transparent 4px) 100% 0,linear-gradient(to top, #55555580 4px, transparent 4px) 0 100%,linear-gradient(to top, #55555580 4px, transparent 4px) 100% 100%;
  background-repeat: no-repeat;
  background-size: 20px 20px;
}

.c0:after {
  content: '';
  position: absolute;
  display: inline-block;
  top: 10%;
  left: 10%;
  right: 10%;
  bottom: 10%;
  opacity: 0.2;
  background: linear-gradient(to right, #55555580 4px, transparent 4px) calc(50% + 30px - 2px) 50%,linear-gradient(to bottom, #55555580 4px, transparent 4px) 50% calc(50% + 30px - 2px);
  background-repeat: no-repeat;
  background-size: 60px 60px;
}

.c1 {
  width: 100%;
  transform: scaleX(-1);
}

<div
  className="c0"
>
  <video
    className="c1"
    muted={true}
  />
  <canvas
    hidden={true}
  />
</div>
`;

exports[`QRReader modified viewfinder color 1`] = `
.c0 {
  width: 100%;
  position: relative;
}

.c0:before {
  content: '';
  position: absolute;
  display: inline-block;
  top: 10%;
  left: 10%;
  right: 10%;
  bottom: 10%;
  opacity: 0.2;
  background: linear-gradient(to right, red 4px, transparent 4px) 0 0,linear-gradient(to right, red 4px, transparent 4px) 0 100%,linear-gradient(to left, red 4px, transparent 4px) 100% 0,linear-gradient(to left, red 4px, transparent 4px) 100% 100%,linear-gradient(to bottom, red 4px, transparent 4px) 0 0,linear-gradient(to bottom, red 4px, transparent 4px) 100% 0,linear-gradient(to top, red 4px, transparent 4px) 0 100%,linear-gradient(to top, red 4px, transparent 4px) 100% 100%;
  background-repeat: no-repeat;
  background-size: 20px 20px;
}

.c0:after {
  content: '';
  position: absolute;
  display: inline-block;
  top: 10%;
  left: 10%;
  right: 10%;
  bottom: 10%;
  opacity: 0.2;
  background: linear-gradient(to right, red 4px, transparent 4px) calc(50% + 30px - 2px) 50%,linear-gradient(to bottom, red 4px, transparent 4px) 50% calc(50% + 30px - 2px);
  background-repeat: no-repeat;
  background-size: 60px 60px;
}

.c1 {
  width: 100%;
}

<div
  className="c0"
>
  <video
    className="c1"
    muted={true}
  />
  <canvas
    hidden={true}
  />
</div>
`;

exports[`QRReader no viewfinder 1`] = `
.c0 {
  width: 100%;
  position: relative;
}

.c1 {
  width: 100%;
}

<div
  className="c0"
>
  <video
    className="c1"
    muted={true}
  />
  <canvas
    hidden={true}
  />
</div>
`;
